### PR TITLE
prov/efa: refactor code in efa_device

### PR DESCRIPTION
**Pull request by wzamazon**
Refactor work on efa_device.c

### CODE REVIEW

File: prov/efa/docs/naming_convention.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+`alloc` has the same behavior as `open`. It allocate memory of an object, then
+initializes its data members. `open` is used for larger object, like endpoint (ep)
+and completion queue (cq). `alloc` is used on smaller object, like `tx_entry`
+and `rx_entry`.
+
+`release` works on the opposite direction of `alloc`, and is used on object
+that `alloc` is used on.
+
+`initialize` is used to initialze global variables. For example, `efa_device_list_initialize`
```
Comment (shijin-aws) [inline prov/efa/docs/naming_convention.md:55]: I do find `fi_ops` uses `init` and `fin` for `initialize` and `finalize`. Do we want to follow that convention?

Comment (wzamazon) [reply prov/efa/docs/naming_convention.md:55]: The common code uses `init` for both normal variables and global variables. However, for EFA provider, I want to make that distinction. So I think keep it makes sense.

File: prov/efa/src/efa_device.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	union ibv_gid		ibv_gid;
+	uint32_t		device_caps;
+	uint32_t		max_rdma_size;
+};
+
+int efa_device_list_initialize(void);
+
+void efa_device_list_finalize(void);
+
+extern struct efa_device *g_device_list;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_device.h:55]: `g_` for global variable. I found it very help to identify a variable as global variable when read code.

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_device.h:55]: Eventually I want all global variable's name start with a `g_` prefix.

[the remark this answers is not in the document]
Comment (shijin-aws) [reply prov/efa/src/efa_device.h:55]: I see, could you add this rule to the naming convention file? Haven't seen it yet.

[the remark this answers is not in the document]
Comment (shijin-aws) [reply prov/efa/src/efa_device.h:55]: I see, could you add this rule to the naming convention file? Haven't seen it yet.

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/efa_device.h:55]: Thank you! Added in new revision.

File: prov/efa/src/efa.h [hunk 3]
```diff
@@ -164,7 +165,7 @@ struct efa_domain {
 	struct util_domain	util_domain;
 	enum efa_domain_type	type;
 	struct fid_domain	*shm_domain;
-	struct efa_context	*ctx;
+	struct efa_device	*device;
```
Comment (shijin-aws) [inline prov/efa/src/efa.h:168]: I find `efa_device` actually replaces both `efa_context` and `efa_device_attr`. Probably mentioning it as well in the commit message. The current one makes me think `efa_device` only replaces `efa_context`.

Comment (wzamazon) [reply prov/efa/src/efa.h:168]: Thank you! Updated commit message in new revision.

**Review by shijin-aws — APPROVED** [review]

**Pull request merged**